consensus/core/src/commit_consumer.rs

```diff
@@ -38,24 +39,66 @@ impl CommitConsumer {
 }
 
 pub struct CommitConsumerMonitor {
-    highest_handled_commit: AtomicU32,
+    // highest commit that has been handled by Sui
+    highest_handled_commit: watch::Sender<u32>,
+
+    // the highest commit found in local storage at startup
+    highest_observed_commit_at_startup: RwLock<u32>,
 }
 
 impl CommitConsumerMonitor {
     pub(crate) fn new(last_handled_commit: CommitIndex) -> Self {
         Self {
-            highest_handled_commit: AtomicU32::new(last_handled_commit),
+            highest_handled_commit: watch::Sender::new(last_handled_commit),
+            highest_observed_commit_at_startup: RwLock::new(0),
         }
     }
 
     pub(crate) fn highest_handled_commit(&self) -> CommitIndex {
-        self.highest_handled_commit
-            .load(std::sync::atomic::Ordering::Acquire)
+        *self.highest_handled_commit.borrow()
     }
 
     pub fn set_highest_handled_commit(&self, highest_handled_commit: CommitIndex) {
         self.highest_handled_commit
-            .store(highest_handled_commit, std::sync::atomic::Ordering::Release);
+            .send(highest_handled_commit)
+            .unwrap();
+    }
+
+    pub fn highest_observed_commit_at_startup(&self) -> CommitIndex {
+        *self.highest_observed_commit_at_startup.read().unwrap()
+    }
+
+    pub fn set_highest_observed_commit_at_startup(
+        &self,
+        highest_observed_commit_at_startup: CommitIndex,
+    ) {
+        let highest_handled_commit = self.highest_handled_commit();
+        assert!(
+            highest_observed_commit_at_startup >= highest_handled_commit,
+            "we cannot have handled a commit that we do not know about: {} < {}",
+            highest_observed_commit_at_startup,
+            highest_handled_commit,
+        );
+
+        let mut commit = self.highest_observed_commit_at_startup.write().unwrap();
+
+        assert!(
+            *commit == 0,
+            "highest_known_commit_at_startup can only be set once"
+        );
+        *commit = highest_observed_commit_at_startup;
+    }
+
+    pub(crate) async fn replay_complete(&self) {
+        let highest_observed_commit_at_startup = self.highest_observed_commit_at_startup();
+        let mut rx = self.highest_handled_commit.subscribe();
+        loop {
+            let highest_handled = *rx.borrow_and_update();
+            if highest_handled >= highest_observed_commit_at_startup {
+                return;
+            }
+            rx.changed().await.unwrap();
```

Comment:
Maybe exit the loop instead of panic in case of shutdown order issues?

Reply:
I think hitting the unwrap is impossible - this method has a `&self` reference, so the watch::Sender cannot possibly be dropped while we are in this method.